### PR TITLE
fix: #6 cursor placement

### PR DESCRIPTION
**Pull request by blackfalcon**
# Alaska Airlines Pull Request

**Fixes:** #6

## Summary:

This update will address the bug related to moving the cursor on iOS safari. 

issue_6

## Type of change:

Please delete options that are not relevant.

- [ ] New capability
- [x] Revision of an existing capability
- [ ] Infrastructure change (automation, etc.)
- [ ] Other (please elaborate)

## Checklist:

- [x] My update follows the CONTRIBUTING guidelines of this project
- [x] I have performed a self-review of my own update

**By submitting this Pull Request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

_Pull Requests will be evaluated by their quality of update and whether it is consistent with the goals and values of this project. Any submission is to be considered a conversation between the submitter and the maintainers of this project and may require changes to your submission._

**Thank you for your submission!**<br>
-- Auro Design System Team

### CODE REVIEW

File: src/base-input.js [hunk 1]
```diff
@@ -146,9 +146,7 @@ export default class BaseInput extends LitElement {
    * @param {string} el String entered into input.
    * @returns {string} Validates string entered into the input field.
    */
-  handleInput(el) {
-    this.value = el.target.value;
```
[the remark this answers is not in the document]
Comment (geoffrich) [reply src/base-input.js:150]: Maybe we implement a getter for the value prop that returns this.inputElement.value? https://lit-element.polymer-project.org/guide/properties#accessors-custom

File: src/auro-input.js [hunk 2]
```diff
@@ -80,7 +78,10 @@ export default class AuroInput extends BaseInput {
    */
   handleKeyUp() {
     const iconContainer = this.shadowRoot.querySelector('.iconContainer'),
-      inputValue = this.shadowRoot.getElementById(this.id).value;
+      inputValue = this.shadowRoot.getElementById(this.id).value,
+      labelElement = this.shadowRoot.querySelector('label');
```
[the remark this answers is not in the document]
Comment (blackfalcon) [reply src/auro-input.js:82]: Good catches ... I got lazy trying to just figure out how to make this work.

File: src/base-input.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (blackfalcon) [reply src/base-input.js:150]: Here is what I just learned. This is not a touch issue, this is a Safari issue as a whole. The way that this.input is being re-written by the input event, it kicks the cursor to the end as the value is updated versus keeping it in the new place as the starting point. 

Getters and Setters (I need help). I kind of understand what it is doing, but no matter what I try, it breaks all the things. 

This is not the proposed solution, but this is where I got stuck. By NOT resetting this.value, the cursor issues is gone. And you can directly query el.inputValue and get a return. But el.value is not set unless set on load. And even then, it is not updated by the input events.

[the remark this answers is not in the document]
Comment (geoffrich) [reply src/base-input.js:159]: I'll try some things out and report back.

[the remark this answers is not in the document]
Comment (geoffrich) [reply src/base-input.js:159]: Adobe's sp-textfield does not have this issue. Looks like they manage the selection range on input. We might need to do something similar: https://github.com/adobe/spectrum-web-components/blob/b0bca7a1b337d73baa3e2df6357a3829a5c625e1/packages/textfield/src/Textfield.ts#L91-L111

[the remark this answers is not in the document]
Comment (blackfalcon) [reply src/base-input.js:159]: You mean this part? https://github.com/adobe/spectrum-web-components/blob/b0bca7a1b337d73baa3e2df6357a3829a5c625e1/packages/textfield/src/Textfield.ts#L95-L103

[the remark this answers is not in the document]
Comment (geoffrich) [reply src/base-input.js:159]: Yeah, that part.

[the remark this answers is not in the document]
Comment (geoffrich) [reply src/base-input.js:159]: Looks like `setSelectionRange` is the key, which has very good browser support. https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/setSelectionRange

[the remark this answers is not in the document]
Comment (blackfalcon) [reply src/base-input.js:159]: I am simply stuck. I tried the following solution and this made things WAY worse. 

```js
  handleInput(el) {
    const inputElement = this.shadowRoot.querySelector('.inputElement');
    let nextSelectStart = el.target.selectionStart - 1;

    this.value = el.target.value;

    inputElement.setSelectionRange (
      nextSelectStart,
      nextSelectStart
    )

    if (this.hasBlurred) {
      this.validate();
    }
  }
```

[the remark this answers is not in the document]
Comment (geoffrich) [reply src/base-input.js:159]: I think I fixed it, but don't currently have a way to test the change on my phone. We were looking at the wrong lines -- we needed the [section outside the if block](https://github.com/adobe/spectrum-web-components/blob/b0bca7a1b337d73baa3e2df6357a3829a5c625e1/packages/textfield/src/Textfield.ts#L106-L110). The code inside the if block was blocking certain keys and resetting the selection.

[the remark this answers is not in the document]
Comment (geoffrich) [reply src/base-input.js:159]: Confirmed that it's fixed running locally on my iPhone.

File: test/auro-input.test.js [hunk 3]
```diff
@@ -113,17 +113,6 @@ describe('auro-input', () => {
     expect(input.name).to.equal('test');
   });
 
-  it('sets value when input event triggered', async () => {
```
[the remark this answers is not in the document]
Comment (geoffrich) [reply test/auro-input.test.js]: I put the test back.

File: src/base-input.js [hunk 4]
```diff
@@ -155,12 +155,17 @@ export default class BaseInput extends LitElement {
    * @private Validates value of the input in blur
    * @returns {string} Validates string entered into the input field.
    */
-  handleInput(el) {
-    this.inputValue = el.target.value;
-
+  handleInput(e) {
```
Comment (blackfalcon) [inline src/base-input.js:158]: Is `e` still needed? This value of this is not being used or being passed in? 

Locally, I removed it and it all seemed to work?

Comment (geoffrich) [reply src/base-input.js]: Yep, it can be removed. Fixed.

**Review by geoffrich — APPROVED** [review]

**Pull request merged**